File: content/LLD/Reference-Code/13. Abstract Factory Pattern/bad_example.py
from abc import ABC, abstractmethod


class Food(ABC):
    @abstractmethod
    def prepare(self):
        pass


# ============= NORTH INDIAN DISHES =============


class PaneerTikka(Food):
    def prepare(self):
        print("Preparing Paneer Tikka (North Indian Starter)")


class ButterChicken(Food):
    def prepare(self):
        print("Preparing Butter Chicken (North Indian Main Course)")


class GulabJamun(Food):
    def prepare(self):
        print("Preparing Gulab Jamun (North Indian Dessert)")


# ============= SOUTH INDIAN DISHES =============


class MeduVada(Food):
    def prepare(self):
        print("Preparing Medu Vada (South Indian Starter)")


class Dosa(Food):
    def prepare(self):
        print("Preparing Dosa (South Indian Main Course)")


class Payasam(Food):
    def prepare(self):
        print("Preparing Payasam (South Indian Dessert)")


# ============= CHINESE DISHES =============


class SpringRolls(Food):
    def prepare(self):
        print("Preparing Spring Rolls (Chinese Starter)")


class FriedRice(Food):
    def prepare(self):
        print("Preparing Fried Rice (Chinese Main Course)")


class FortuneCookie(Food):
    def prepare(self):
        print("Preparing Fortune Cookie (Chinese Dessert)")


class RestaurantService:
    def create_meal(self, cuisine_type: str):
        if cuisine_type == "north_indian":
            starter = PaneerTikka()
            main_course = ButterChicken()
            dessert = GulabJamun()

        elif cuisine_type == "south_indian":
            starter = MeduVada()
            main_course = Dosa()
            dessert = Payasam()

        elif cuisine_type == "chinese":
            starter = SpringRolls()
            main_course = FriedRice()
            dessert = FortuneCookie()

        else:
            print("Cuisine not available!")
            return None

        # Prepare the meal
        starter.prepare()
        main_course.prepare()
        dessert.prepare()


restaurant = RestaurantService()
restaurant.create_meal("north_indian")

File: content/LLD/Reference-Code/13. Abstract Factory Pattern/BadExample.java
public class BadExample {
    interface Food {
        void prepare();
    }

    // North Indian
    static class PaneerTikka implements Food {
        public void prepare() { System.out.println("Preparing Paneer Tikka (North Indian Starter)"); }
    }
    static class ButterChicken implements Food {
        public void prepare() { System.out.println("Preparing Butter Chicken (North Indian Main Course)"); }
    }
    static class GulabJamun implements Food {
        public void prepare() { System.out.println("Preparing Gulab Jamun (North Indian Dessert)"); }
    }

    // South Indian
    static class MeduVada implements Food {
        public void prepare() { System.out.println("Preparing Medu Vada (South Indian Starter)"); }
    }
    static class Dosa implements Food {
        public void prepare() { System.out.println("Preparing Dosa (South Indian Main Course)"); }
    }
    static class Payasam implements Food {
        public void prepare() { System.out.println("Preparing Payasam (South Indian Dessert)"); }
    }

    // Chinese
    static class SpringRolls implements Food {
        public void prepare() { System.out.println("Preparing Spring Rolls (Chinese Starter)"); }
    }
    static class FriedRice implements Food {
        public void prepare() { System.out.println("Preparing Fried Rice (Chinese Main Course)"); }
    }
    static class FortuneCookie implements Food {
        public void prepare() { System.out.println("Preparing Fortune Cookie (Chinese Dessert)"); }
    }

    static class RestaurantService {
        public void createMeal(String cuisineType) {
            Food starter, mainCourse, dessert;
            if ("north_indian".equalsIgnoreCase(cuisineType)) {
                starter = new PaneerTikka();
                mainCourse = new ButterChicken();
                dessert = new GulabJamun();
            } else if ("south_indian".equalsIgnoreCase(cuisineType)) {
                starter = new MeduVada();
                mainCourse = new Dosa();
                dessert = new Payasam();
            } else if ("chinese".equalsIgnoreCase(cuisineType)) {
                starter = new SpringRolls();
                mainCourse = new FriedRice();
                dessert = new FortuneCookie();
            } else {
                System.out.println("Cuisine not available!");
                return;
            }

            starter.prepare();
            mainCourse.prepare();
            dessert.prepare();
        }
    }

    public static void main(String[] args) {
        RestaurantService restaurant = new RestaurantService();
        restaurant.createMeal("north_indian");
    }
}
